_posts/2020-12-01-the-road-to-the-jit.md

```diff
@@ -0,0 +1,382 @@
+  be nice for [Wings3D][wings3d] to have floating point literals.  The
+  OTP team implemented literal pools in OTP R12.
+
+[dialyzer]: http://erlang.org/doc/apps/dialyzer/index.html
+[newtags]: http://www.it.uu.se/research/publications/reports/2000-029/2000-029-nc.pdf
+[bitstrings]: http://user.it.uu.se/%7Epergu/papers/erlang05.pdf
+[core]: https://www.it.uu.se/research/group/hipe/cerl/doc/core_erlang-1.0.3.pdf
+[core_introduction]: http://www.erlang.se/workshop/carlsson.ps
+[core_by_example]: http://blog.erlang.org/core-erlang-by-example
+[exception]: https://erlang.org/workshop/2004/exception.pdf
+[richcarl]: https://github.com/richcarl
+[wings3d]: http://www.wings3d.com
+
+### The Tracing JIT projects
+
+There have been three separate research projects that tried to develop
+a tracing JIT for Erlang.
+
+A tracing JIT (Just In Time compiler) is a JIT that runs in two phases:
+
+* First it traces execution to find sequences of hot (frequently executed)
+  code.
+
+* It then rewrites the found traces to native code.
+
+The goals for the three JIT projects were:
+
+* The JIT should work automatically with no need for the user to identify
+  which modules to compile to native code beforehand.
+
+* There should be total feature compatibility with the non-JIT
+  BEAM. In particular, tracing, scheduling behavior, save calls, and
+  hot code reloading should continue to work, and stack traces should
+  be identical to the ones in the non-JIT BEAM.
+
+* The system should at least on average never be slower than the non-JIT
+  BEAM.
+
+There were some promising results when running some benchmarks, but ultimately
+it turned out to be impossible to fulfill the goal to never be slower than the
+non-JIT system.  Here are the main reasons for the slowdowns:
+
+* To be able to do the tracing (finding hot code), the BEAM interpreter needed
+  tweaking.  It was difficult to be able to do tracing without lowering the
+  base speed of the BEAM interpreter.
+
+* It was also difficult to design the mechanism for context
+  switching between the interpreted code and native code in a way that
+  didn't lower the base speed of the BEAM interpreter.
+
+* When a hot sequence of code has been found, the code needed to be compiled
+  to native code.  The compilation, that used LLVM, was slow.
+
+* When a hot sequence had finally been converted to native code, it could turn
+  out that it would not be executed again.  That was particularly a problem for
+  the Erlang compiler that runs many passes.  Typically, when some of the code
+  for one pass had been converted to native code, the compiler was already running
+  the next pass.
+
+The later projects mitigated some of the issues in the previous projects.  For example,
+the compilation was reduced by doing more optimizations before invoking LLVM.
+Ultimately, though, it was decided to terminate the third and final tracing JIT project
+at the end of 2019.
+
```

Comment:
I think a link to this paper could be useful:

https://dl.acm.org/doi/abs/10.1145/2633448.2633450?casa_token=tp5GAPlARNcAAAAA%3AEQbE8U2rzCpMKrkzE2eQbbS_fgKuKs7sAuvHFPWO3RjkU1p-3XsMi-Pkbc8A1DFurvvsa6t7qRuI

Does it describe the first attempt?

Perhaps one could find links to talks or similar describing the other two attempts?